README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# cjbart

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/cjbart)](https://CRAN.R-project.org/package=cjbart)
[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/tsrobinson/cjbart/workflows/R-CMD-check/badge.svg)](https://github.com/tsrobinson/cjbart/actions)
[![R-CMD-check](https://github.com/tsrobinson/cjbart/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/tsrobinson/cjbart/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

## Overview

**cjbart** is an R package for analyzing conjoint experiments using
Bayesian Additive Regression Trees (BART), specifically focusing on
inspecting heterogeneous treatment effects.

This package is in its early stages of development, and core
functionality is liable to change.

## Installation

The latest development version of **cjbart** can be installed directly
from this repository, using the following code:

``` r
# install.packages("devtools")
devtools::install_github("tsrobinson/cjbart")
```

## Getting help

If you come across any issues, or have any suggestions for improvements,
please raise an issue
[here](https://github.com/tsrobinson/cjbart/issues).


NEWS.md
# cjbart 0.3.2

* Update documentation: correct details line for `cjbart()`.
* Add compatibility so passing seed argument on Windows returns warning rather than error
* Add unit tests for OS compatibility

# cjbart 0.3.1

* Fix documentation: to plot the variable importance grid, the `x` argument takes the result of calling `het_vimp()` and not `IMCE()` as previously documented (in v0.3.0)

* Update package dependencies

# cjbart 0.3.0

* Rename arguments in `het_vimp()` from `model` to `imces` and `outcomes` to `levels` for consistency across functions
* Change `het_vimp()` so it returns a list, including the results and an attribute-level look-up table
* Added ability to run `het_vimp()` calculation in parallel
* Fixed bug when predicting continuous conjoint outcomes
* Add new `pIMCE()` function to calculate population individual-level marginal component effects
* Add new `AMCE()` function that includes Bayesian uncertainty estimate
* Add check for unique levels across attributes. If not, the attribute-level values are modified to ensure uniqueness
* Add functionality to improve quality of outputs
* Update the vignette to reflect new functionality
* Add ability to drop attributes from IMCE analysis
* Remove dependencies not needed

# cjbart 0.2.2

* Re-release after dependencies back on CRAN

# cjbart 0.2.1

* Minor internal changes to keep pace with R development
* Minor changes to vignette wording

# cjbart 0.2.0

* Addresses unbalanced chunk delimiters in package vignette
* Adds new functionality for analyzing heterogeneity after IMCE prediction using random forest variable importance measures

# cjbart 0.1.0

* First release including all core functionality
* Includes wrapper of BART estimation, IMCE prediction, and generics


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(cjbart)

test_check("cjbart")


tests/testthat/test-check_windows.R
test_that("runs model with seed warning on Windows", {

  test_df <- data.frame(Y = rbinom(50,1, prob = 0.5),
                        X1 = runif(50),
                        X2 = runif(50),
                        id = 1:50)

  if (.Platform$OS.type=='unix') {
    testthat::expect_no_warning(
      cjbart::cjbart(data = test_df, Y = "Y", id = "id", seed = 1, cores = 1,
                     ntree = 5, numcut = 10)
    )
  } else {
    testthat::expect_warning(
      cjbart::cjbart(data = test_df, Y = "Y", id = "id", seed = 1, cores = 1,
                     ntree = 5, numcut = 10)
    )
  }

})


R/pIMCE.R
#' Population-Weighted Heterogeneous Effects Analysis of Conjoint Results
#'
#' @description \code{pIMCE} calculates the population individual-level marginal component effects from a BART-estimated conjoint model, using marginal attribute distributions specified by the researcher.
#' @param model A model object, the result of running \code{cjbart()}
#' @param covar_data A data.frame of covariate information to predict pIMCEs over
#' @param attribs Vector of attribute names
#' @param l Name of the attribute of interest
#' @param l_1 Attribute-level of interest for attribute *l*
#' @param l_0 Reference level for attribute *l*
#' @param marginals A named list where every element is a named vector of marginal probabilities for each corresponding attribute-level.  For example, \code{marginals = list("A1" = c("q" = 0.4, "r" = 0.6), "A2" = c("x" = 0.7, "y" = 0.2, "z" = 0.1))}
#' @param method Character string, setting the variance estimation method to use. When method is "parametric", a typical combined variance estimate is employed; when \code{method = "bayes"}, the 95% posterior interval is calculated; and when \code{method = "rubin"}, combination rules are used to combine the variance analogous to in multiple imputation analysis.
#' @param alpha Number between 0 and 1 -- the significance level used to compute confidence/posterior intervals. When \code{method = "bayes"}, the posterior interval is calculated by taking the alpha/2 and (1-alpha/2) quantiles of the posterior draws. When \code{method = "rubin"}, the confidence interval equals the IMCE +/- \code{qnorm(alpha/2)}. By default, alpha is 0.05 i.e. generating a 95% confidence/posterior interval.
#' @param cores Number of CPU cores used during prediction phase
#' @param skip_checks Boolean, indicating whether to check the structure of the data (default = \code{FALSE}). Only set this to \code{TRUE} if you are confident that the data is structured appropriately
#' @param verbose Boolean, indicating whether to print progress (default = TRUE)
#' @details This function calculates the population-weighted IMCE, which takes into account the population distribution of profiles. Rather than average over the multiple OMCE estimates, this function generates estimated treatment effects for *all* possible potential outcomes along all attributes except the attribute of interest, and then marginalizes these over the supplied marginal distributions. Uncertainty estimates are recovered using credible intervals.
#' @return \code{pIMCE} returns a data.frame of population-weighted estimates, credible interval bounds, and the covariate information supplied
#' @seealso [cjbart::cjbart()]
#' @importFrom stats predict
#' @export
pIMCE <- function(model,
                  covar_data,
                  attribs,
                  l,
                  l_1,
                  l_0,
                  marginals,
                  method = "bayes",
                  alpha = 0.05,
                  cores = 1,
                  skip_checks = FALSE,
                  verbose = TRUE) {

  # Get variables from trained model
  Y <- model$Y_col
  round <- model$round_col
  id_var <- model$id_col
  type <- model$type

  # Check optional args
  if (!(method %in% c("average","bayes","rubin"))) {
    stop("Variance estimation method must be in c('parametric','bayes','rubin'). See ?OMCE for more details.")
  }

  if (!skip_checks) {

    # Check marginals
    for (attrib in attribs[attribs != l]) {
      if (!exists(attrib, where= marginals)) {

        unq_levels <- model$factor_levels[[attrib]]
        unq_levels_n <- length(unq_levels)
        marginals[[attrib]] <- rep(1/unq_levels_n,unq_levels_n)
        names(marginals[[attrib]]) <- unq_levels

        warning(paste0("Marginal probabilities not provided for attribute ", attrib,". Assuming a uniform distribution."))
      } else if (sum(marginals[[attrib]]) != 1) {
        stop(paste0("Marginal probabilities must sum to 1. First error found for attribute ", attrib))
      }
    }

  }

  for (v in names(model$factor_levels)) {
    if (v %in% colnames(covar_data)) {
      covar_data[[v]] <- factor(covar_data[[v]], levels = model$factor_levels[[v]])
    }
  }

  if (id_var %in% names(covar_data)) {
    id_vec <- covar_data[[id_var]]
    covar_data[[id_var]] <- NULL
  } else {
    id_vec <- NULL
  }

  # Generate potential outcomes for L-1 attributes
  attribs_res <- attribs[attribs != l]
  levels_res <- model$factor_levels[attribs_res]
  pot_outs_res <- expand.grid(levels_res)

  for (v in names(pot_outs_res)) {
    pot_outs_res[[v]] <- factor(pot_outs_res[[v]], levels = model$factor_levels[[v]])
  }

  tau_wgts <- apply(pot_outs_res, 1, function(x) {
    prod(sapply(attribs[attribs != l], function (y) marginals[[y]][[x[[y]]]]))
  })

  preds <- sapply(1:nrow(covar_data), function (i) {

    if (verbose) {
      cat(paste0("\rPredicting covariate profile ",i,"/",nrow(covar_data),""))
    }

    pred_data <- cbind(pot_outs_res,
                       covar_data[i,], row.names = NULL)

    phats <- list()
    for (l_d in c(l_1, l_0)) {

      pred_data[[l]] <- factor(l_d, levels = model$factor_levels[[l]])

      if (type == "choice") {
        phats[[l_d]] <- .quiet(predict(
          model,
          newdata = BART::bartModelMatrix(pred_data)[,names(model$varcount.mean)],
          mc.cores = cores
        )$prob.test)
      } else {
        phats[[l_d]] <- .quiet(predict(
          model,
          newdata = BART::bartModelMatrix(pred_data)[,names(model$varcount.mean)],
          mc.cores = cores
        )$yhat.test)
      }
    }

    omce_mat <- phats[[l_1]] - phats[[l_0]]
    pIMCE_z <- omce_mat%*%tau_wgts
    pIMCE_est <- mean(pIMCE_z)
    pIMCE_lower <- stats::quantile(pIMCE_z, alpha/2)[[1]]
    pIMCE_upper <- stats::quantile(pIMCE_z, 1-alpha/2)[[1]]

    return(c(pIMCE = pIMCE_est, pIMCE_lower = pIMCE_lower, pIMCE_upper = pIMCE_upper))

  })

  if (verbose) {
    cat("\n")
  }

  if (!is.null(id_vec)) {
    covar_data[[id_var]] <- id_vec
  }

  return(cbind(t(preds), covar_data))
}



R/utils.R
# Convert character columns to factors with warning
.char_to_fact <- function(data) {

  missing_factors <- sapply(data, is.character)

  if (sum(missing_factors) > 0) {

    missing_vars <- names(missing_factors)[missing_factors]

    data[,missing_vars] <- Map(as.factor, subset(data, select = missing_vars))

    # message("The following variables were converted to factors: ",
    #         paste0(missing_vars, collapse = ", ")
    # )

  }

  return(data)
}

# Suppress console output where verbose option not present
.quiet <- function(x) {

  # Code from Hadley Wickham
  # See: https://r.789695.n4.nabble.com/Suppressing-output-e-g-from-cat-td859876.html

  sink(tempfile())
  on.exit(sink())
  invisible(force(x))

}


# Implementation of Rubin's combination rules
.combine <- function(theta, var_theta) {

  m <- length(theta)

  Q_bar <- (1/m)*sum(theta)
  U_bar <- (1/m)*sum(var_theta)

  demean <- (theta-Q_bar)^2

  B <- (1/(m-1)) * sum(demean)

  Q_bar_var <- U_bar + (1 + (1/m))*B
  Q_bar_se <- sqrt(Q_bar_var)

  v_m <- (m-1)*(1+(U_bar/((1+m^-1)*B)))^2

  std_err = Q_bar_se

  return(std_err)

}


R/generics.R
#' Plot Marginal Component Effects of a \code{cjbart} Object
#' @description Plots observation-level or individual-level marginal component effects (OMCE and IMCE respectively). By default, all attribute-levels in the model are plotted.
#' @param x Object of class \code{cjbart}, the result of running [cjbart::IMCE()]
#' @param covar Character string detailing the covariate over which to analyze heterogeneous effects
#' @param plot_levels Optional vector of conjoint attribute levels to plot. If not supplied, all attributes within the conjoint model will be plotted.
#' @param se Boolean determining whether to show an estimated 95% confidence interval
#' @param ... Additional arguments for plotting the marginal component effects (see below).
#' @return Plot of marginal component effects.
#' @importFrom rlang .data
#' @method plot cjbart
#' @export
plot.cjbart <- function(x, covar = NULL, plot_levels = NULL, se = TRUE,  ...) {

  exp_data <- x$imce

  plot_data <- tidyr::pivot_longer(cols = x$att_levels,
                                   names_to = "level",
                                   values_to = "imce",
                                   data = exp_data)

  plot_data <- merge(x = plot_data, y = x$att_lookup,
                     all.x = TRUE,
                     by = "level")

  if (se) {

    ci_lower <- tidyr::pivot_longer(cols = x$att_levels,
                                    names_to = "level",
                                    values_to = "imce_lower",
                                    data = x$imce_lower)

    ci_upper <- tidyr::pivot_longer(cols = x$att_levels,
                                    names_to = "level",
                                    values_to = "imce_upper",
                                    data = x$imce_upper)

    plot_data <- merge(x = plot_data,
                       y = ci_lower[,c(x$id,"level","imce_lower")],
                       all.x = TRUE,
                       by = c("level",x$id))

    plot_data <- merge(x = plot_data,
                       y = ci_upper[,c(x$id,"level","imce_upper")],
                       all.x = TRUE,
                       by = c("level",x$id))
  }

  if (!is.null(plot_levels)) {
    plot_data <- plot_data[plot_data$level %in% plot_levels,]

    if (nrow(plot_data) == 0) {
      stop("Failed to find attribute plot_levels in data. This may be because the conjoint has multiple attributes with the same level.\n If this is the case, try specifying the required levels as '<attribute>_level', e.g., '<age>_80+'.")
    }

  }

  plot_data <- do.call(

    rbind,

    by(
      plot_data,
      INDICES = plot_data$level,
      FUN = function(x) {
        x_ordered <- x[order(x$imce),]
        x_ordered$x_order <- 1:nrow(x_ordered)
        return(x_ordered)
        }
     )
  )

  plot_data$facet_labels <- paste0("atop(italic('",plot_data$Attribute,":'),bold('",plot_data$Level,"'))")

  base_plot <- ggplot2::ggplot(plot_data,
                                 ggplot2::aes_string(x = "x_order",
                                                     y = "imce")) +

    ggplot2::geom_hline(yintercept = 0, size = 0.5, linetype = "dashed") +

    {if (!is.null(covar)) {ggplot2::aes_string(color = covar)}} +
    {if (se) {ggplot2::geom_ribbon(ggplot2::aes_string(ymin = "imce_lower", ymax = "imce_upper"), color = NA, fill = "grey60", alpha = 0.7)}} +

    ggplot2::facet_wrap(~.data$facet_labels, scales = "free", labeller = "label_parsed") +
    ggplot2::geom_point(alpha = 0.8) +
    ggplot2::ylab("IMCE") +
    ggplot2::xlab("Individual") +
    ggplot2::theme(legend.position = "bottom",
                   axis.text.x=ggplot2::element_blank(),
                   axis.ticks.x=ggplot2::element_blank())

  if (!is.null(covar)) {

    if (typeof(plot_data[[covar]]) == "double") {

      base_plot <- base_plot +
        ggplot2::scale_color_gradient(low = "#0072B2", high = "#E69F00")

    } else {

      base_plot <- base_plot +
        ggplot2::scale_colour_manual(values=c("#000000", # Colour-blind friendly pallete
                                              "#E69F00",
                                              "#56B4E9",
                                              "#009E73",
                                              "#F0E442",
                                              "#0072B2",
                                              "#D55E00",
                                              "#CC79A7"))
    }

  }


  return(base_plot)
}

#' Summarizing \code{cjbart} Marginal Component Effect Estimates
#' @description \code{summary} method for class "cjbart"
#' @param object Object of class \code{cjbart}, the result of running [cjbart::IMCE()]
#' @param ... Further arguments (not currently used)
#' @return Data frame summarizing the average marginal component effect (AMCE), the minimum and maximum values, and standard deviations for each attribute-level.
#' @note To calculate the AMCE with Bayesian credible intervals, please use the \code{AMCE()} function instead.
#' @method summary cjbart
#' @seealso [cjbart::AMCE()]
#' @export
summary.cjbart <- function(object, ...) {

  # IMCE summary
  IMCE_only <- subset(object$imce, select = object$att_levels)

  AMCE <- colMeans(IMCE_only)
  mins <- apply(IMCE_only,2,min)
  maxs <- apply(IMCE_only,2,max)
  sds <- apply(IMCE_only,2,stats::sd)

  summary_tab <- data.frame(level = object$att_levels,
                            AMCE = AMCE,
                            `Min.` = mins,
                            `Max.` = maxs,
                            `Std.Dev` = sds,
                            row.names = NULL)

  summary_tab <- merge(x = object$att_lookup,
                       y = summary_tab,
                       all.y = TRUE,
                       by = "level")

  summary_tab$level <- NULL

  message("Individual marginal component effects (IMCEs)")
  summary_tab <- tidyr::tibble(summary_tab)
  return(summary_tab)

}

#' Plot Variable Importance Matrix for Heterogeneity Analysis
#' @description Plots a heatmap of variable importance, across predicted IMCEs. By default, all attribute-levels and covariates in the model are plotted.
#' @param x Object of class \code{cjbart.vimp}, the result of running [cjbart::het_vimp()]
#' @param covars Optional vector of covariate names to plot. By default, all included covariates are shown.
#' @param att_levels Optional vector of attribute-levels to plot. By default, all attribute-levels are shown.
#' @param ... Additional arguments (not currently used)
#' @return Plot of covariate importance scores
#' @method plot cjbart.vimp
#' @export
plot.cjbart.vimp <- function(x, covars = NULL, att_levels = NULL,  ...) {

  plot_data <- x$results

  if (!is.null(covars)) {

    plot_data <- plot_data[plot_data$covar %in% covars,]

  }

  if (!is.null(att_levels)) {

    plot_data <- merge(x = plot_data,
                       y = x$att_lookup,
                       by = c("Attribute","Level"),
                       all.x = TRUE)

    plot_data <- plot_data[plot_data$level %in% att_levels,]

  }

  if (is.null(att_levels) | length(att_levels) > 1) {

    ggplot2::ggplot(plot_data,
                    ggplot2::aes_string(x = "covar",
                                        y = "Level",
                                        fill = "importance")) +
      ggplot2::facet_grid(Attribute ~ ., space = "free", scales = "free") +
      ggplot2::geom_tile() +
      ggplot2::scale_fill_gradient(low="white", high="firebrick1") +
      ggplot2::labs(x = "Covariates", y = "Attribute-level", fill = "Importance") +
      ggplot2::theme(axis.text.x = ggplot2::element_text(angle=45, vjust = 1, hjust = 1)) +
      ggplot2::theme_classic()

  } else {

    message("Plotting importance scores for single attribute-level (with 95 percent confidence intervals)")

    ggplot2::ggplot(plot_data,
                    ggplot2::aes_string(y = "covar",
                                        x = "importance",
                                        xmin = "lower2.5",
                                        xmax = "upper97.5")) +
      ggplot2::geom_point() +
      ggplot2::geom_errorbarh(height = 0.2) +
      ggplot2::labs(x = "Standardised Importance",
                    y = "Covariates") +
      ggplot2::xlim(min(plot_data$lower2.5)-1, max(plot_data$upper97.5)+1) +
      ggplot2::labs(title = paste0(unique(plot_data$Attribute),": ", unique(plot_data$Level))) +
      ggplot2::theme_classic()

  }

}



inst/examples/basic_workflow.R
set.seed(89)

subjects <- 500
rounds <- 5
obs <- subjects*rounds

fake_data <- data.frame(A = sample(c("a1","a2","a3"), obs, replace = TRUE),
                        B = sample(c("b1","b2","b3"), obs, replace = TRUE),
                        C = sample(c("c1","c2","c3"), obs, replace = TRUE),
                        D = sample(c("d1","d2","d3"), obs, replace = TRUE),
                        E = sample(c("e1","e2","e3"), obs, replace = TRUE),

                        covar1 = rep(runif(subjects, 0 ,1),
                                     each = rounds),
                        covar2 = rep(sample(c(1,0),
                                            subjects,
                                            replace = TRUE),
                                     each = rounds),

                        id1 = rep(1:subjects, each=rounds),
                        stringsAsFactors = TRUE)

fake_data$Y <- ifelse(fake_data$E == "e2",
                      rbinom(obs, 1, fake_data$covar1),
                      sample(c(0,1), obs, replace = TRUE))

cj_model <- cjbart(data = fake_data,
                   Y = "Y",
                   id = "id1")

het_effects <- IMCE(data = fake_data,
                    model = cj_model,
                    attribs = c("A","B","C","D","E"),
                    ref_levels = c("a1","b1","c1","d1","e1"),
                    Y = "Y",
                    id = "id1",
                    cores = 8)

summary(het_effects)
plot(het_effects, covar = "covar1")


R/vimp.R
#' Estimate Variable Importance Metrics for \code{cjbart} Object
#' @description Estimates random forest variable importance scores for multiple attribute-levels of a conjoint experiment.
#' @param imces Object of class \code{cjbart}, the result of running [cjbart::IMCE()]
#' @param levels An optional vector of attribute-levels to generate importance metrics for. By default, all attribute-levels are analyzed.
#' @param covars An optional vector of covariates to include in the importance metric check. By default, all covariates are included in each importance model.
#' @param cores Number of CPU cores used during VIMP estimation. Each extra core will result in greater memory consumption. Assigning more cores than outcomes will not further boost performance.
#' @param ... Extra arguments (used to check for deprecated argument names)
#' @return A "long" data.frame of variable importance scores for each combination of covariates and attribute-levels, as well as the estimated 95% confidence intervals for each metric.
#' @details Having generated a schedule of individual-level marginal component effect estimates, this function fits a random forest model for each attribute-level using the supplied covariates as predictors. It then calculates a variable importance measure (VIMP) for each covariate. The VIMP method assesses how important each covariate is in terms of partitioning the predicted individual-level effects distribution, and can thus be used as an indicator of which variables drive heterogeneity in the IMCEs.
#'
#' To recover a VIMP measure, we used permutation-based importance metrics recovered from random forest models estimated using [randomForestSRC::rfsrc()]. To permute the data, this function uses random node assignment, whereby cases are randomly assigned to a daughter node whenever a tree splits on the target variable \insertCite{@see @ishwaran2008random}{cjbart}. Importance is defined in terms of how random node assignment degrades the performance of the forest. Higher degradation indicates a variable is more important to prediction.
#'
#' Variance estimates of each variable's importance are subsequently recovered using the delete-d jackknife estimator developed by \insertCite{ishwaran2019standard;textual}{cjbart}. The jackknife method has inherent bias correction properties, making it particularly effective for variable selection exercises such as identifying drivers of heterogeneity.
#' @references \insertAllCited{}
#' @seealso [randomForestSRC::rfsrc()] and [randomForestSRC::subsample()]
#' @importFrom Rdpack reprompt
#' @export
het_vimp <- function(imces, levels = NULL, covars = NULL, cores = 1, ...) {

  if (is.null(levels)) {
    levels <- imces$att_levels
  }

  extra_args = names(list(...))
  deprecated_args = c("model","outcomes")
  deprec_present <- deprecated_args[deprecated_args %in% extra_args]

  if (length(deprec_present) > 0) {
    warning("\nDetected deprecated argument names!\n The following argument names were changed in cjbart v0.3:",
            "\n * `model` (old) -> 'imces' (new) \n * `outcomes` -> `levels`",
            "\n Please check any output and update your scripts accordingly.")
  }

  if (cores == 1 | .Platform$OS.type != "unix") {

    if (cores > 1) {
      warning("Parallelization is not available on your operating system. Reverting to single-core computation.")
    }

    vimps <- lapply(levels,
                    function (x) rf_vimp(model = imces, outcome = x, covars = covars))
  } else {

    vimps <- parallel::mclapply(levels,
                      function (x) rf_vimp(model = imces, outcome = x, covars = covars),
                      mc.cores = cores)

  }

  full_results <- do.call("rbind", vimps)

  full_results <- merge(x = imces$att_lookup,
                        y = full_results,
                        all.y = TRUE,
                        by.x = "level", by.y = "outcome")

  full_results$level <- NULL

  out_obj <- list(results = full_results,
                  att_lookup = imces$att_lookup)

  class(out_obj) <- c("cjbart.vimp")

  return(out_obj)
}

#' Estimate a Single Variable Importance Metric for \code{cjbart} Object
#' @description Estimates random forest variable importance scores for a single attribute-level of a conjoint experiment. This function is for advanced use. Users should typically use the [cjbart::het_vimp()] function.
#' @param model Object of class \code{cjbart}, the result of running [cjbart::IMCE()]
#' @param outcome Character string detailing the covariate over which to analyze heterogeneous effects
#' @param covars An optional vector of covariates to include in the importance metric check. When \code{covars = NULL} (the default), all covariates are included in the importance model.
#' @return Data.frame of variable importance scores for each covariate in the model, as well as values for the estimated 95% confidence interval for each importance score.
#' @export
rf_vimp <- function(model, outcome, covars = NULL) {

  trial_data <- model$imce
  trial_data$outcome <- trial_data[[outcome]]

  if (is.null(covars)) {

    covars <- colnames(model$imce)[!(colnames(model$imce) %in% c(model$att_levels, model$id))]

  }

  trial_data <- trial_data[,c("outcome", covars)]

  trial_data <- stats::na.omit(trial_data)

  n_na <- nrow(model$imce) - nrow(trial_data)

  if (n_na != 0) {
    warning(paste0(n_na," observations removed due to missing covariate data in the model."))
  }

  for (i in 1:length(names(trial_data))) {
    if (typeof(trial_data[,i]) == "character") {
      trial_data[,i] <- as.factor(trial_data[,i])
    }
  }

  message(paste0("Calculating importance metrics for attribute-level: ",outcome))
  rf_mod <- randomForestSRC::rfsrc(outcome ~ ., data = trial_data, importance = "permute")
  vimp_ci <- randomForestSRC::extract.subsample(randomForestSRC::subsample(rf_mod), raw = TRUE)$ci.jk.Z

  att_results <- data.frame(outcome = outcome,
                            covar = colnames(vimp_ci),
                            importance = vimp_ci["50%",],
                            lower2.5 = vimp_ci["2.5%",],
                            upper97.5 = vimp_ci["97.5%",])

  class(att_results) <- c("cjbart.vimp","data.frame")

  return(att_results)

}
